### ISSUE TITLE
metric: AggHistogram windows don't tick when TSDB scrapes metrics registries

### ISSUE THREAD

**Issue by abarganier** [cockroachdb/cockroach#114175]
Relates to: https://github.com/cockroachdb/cockroach/issues/112947, https://github.com/cockroachdb/cockroach/issues/100630

**Describe the problem**

### Some background on histograms & quantile calculations in CockroachDB 

In Prometheus, histogram buckets are scraped in their entirety at the given scrape interval. The diff between bucket counts at each scrape interval is considered a "window", against which range queries using functions like `histogram_quantile` can perform quantile calculations over periods of time (e.g. P90, P99, etc).

In CockroachDB, this is true for the histogram metrics that are exported via the Prometheus-compatible `/_status/vars` endpoint. However, histogram metrics are also written to CockroachDB's internal timeseries database (TSDB). The process of scraping & writing metrics to TSDB is different. The primary difference is that TSDB does *not* store all histogram buckets in their entirety. Instead, at each scrape interval, relevant quantiles (e.g. P99) are calculated for that interval, and the quantiles instead are saved, each as its own metric (e.g. `changefeed.commit_latency-p50`, `changefeed.commit_latency-p99`). 

Since TSDB doesn't store the raw histogram buckets for each scrape interval, and calculates quantiles at scrape time, this means that there's no on-disk representation of a histogram's previous "window" to diff against. Such a previous window is required in order to calculate quantiles for a given interval, though. Therefore, what CockroachDB does is maintain the *previous* histogram window in-memory, so that the current window can be diffed against it to calculate quantiles. 

Eventually the histogram "ticks", where the current window is set to the previous window, and the current window is set to a new histogram with no bucket counts. This "ticking" is normally triggered when the TSDB scrape of the metric registry "inspects" the metric, as we can see here: 
https://github.com/cockroachdb/cockroach/blob/9fdc79b95dcff0cc85cf8b3c8bb21431ce97b8fd/pkg/util/metric/metric.go#L446-L451

If the histogram *didn't* tick, then the values would be cumulative for the lifespan of the CockroachDB node. With a cumulative histogram, and no "ticking", there would be no previous histogram window to calculate quantiles against. What this would look like in practice, is a histogram whose quantile values "level-out" over time, since the diff is always taken against a previous histogram whose counts are all `0`. This would lead to each quantile calculation representing said quantile for the entire lifespan of the node (hardly a useful metric in a **time**series database).

This is why the "windowing" logic within CockroachDB's histogram library is so important. While it has nothing to do with the Prometheus-compatible metrics exported by `/_status/vars`, it has *everything* to do with CockroachDB's internal TSDB. 

### AggHistogram vs. Regular Histograms

In CockroachDB, we have `Histogram` and `AggHistogram`. `Histogram` is the library powering *most* of the histogram metrics in CockroachDB. However, a few metrics with additional levels of cardinality (e.g. changefeeds) use the `AggHistogram` library instead. `AggHistogram` is essentially a wrapper around `Histogram`. The internals are mostly the same. The difference is that an `AggHistogram` represents a *set* of histograms, with one "parent" and multiple "child" histograms. The parent Histogram's counts are equal the sum of all its children. 

One example is if you have multiple changefeeds, where you want to track individual latencies of each changefeed, you can use an `AggHistogram` to give each individual changefeed (the "children") its own histogram, and then use the parent histogram to view the latency across *all* changefeeds. This is what's done with the metric `changefeed.commit_latency`.

### The Problem With AggHistogram

Unfortunately, AggHistograms are bugged. Basically, the "ticking" functionality discussed above was never implemented! Therefore, all `AggHistogram`-powered metrics in CRDB are cumulative for the lifespan of the CRDB node. This leads to quantile metrics "leveling out", as expected (like we discussed above).

For example:
<img width="1270" alt="Screenshot 2023-11-09 at 1 28 06 PM" src="https://github.com/cockroachdb/cockroach/assets/8194877/8c831fa2-ad55-420b-bebd-6a774cd6bfcc">

**Expected behavior**
The `AggHistogram` should "tick" when inspected by TSDB scrapes, just like regular histograms do.

The one potential challenge here is that the "parent" histogram, and all its "children", all need to tick atomically. One thought is to only call `maybeTick` when the parent is inspected, and have the parent be responsible for ticking all its children. This might require the parent and all its children to synchronize on a single mutex, though, which could introduce contention. We should consider this when coming up with a ticking strategy for these AggHistograms.

**TL;DR:**
`AggHistogram`s don't "tick", leading to cumulative quantiles for the lifespan of the node as served by TSDB. Prometheus metrics via `/_status/vars` are not impacted & work correctly. To fix the bug, implement ticking functionality on the `AggHistogram` in a way such that the parent & children histograms all "tick" atomically.

Jira issue: CRDB-33381
